File: Chapter04/Exercise24_04/Exercise24_04.scala
package Exercise24_04

import Utilities01.HelperScala.createSession

object Exercise24_04 {
  def main(args: Array[String]): Unit = {
    val session = createSession(4, "Memory Limits")
    val numbersRDD = session.sparkContext.range(0, 10)
    val mappedNumbersRDD = numbersRDD.mapPartitionsWithIndex((index, partition) => {
      println("@@ Starting with list creation at partition " + index)
      val tenMillionList = List.fill(10000000)(-1)
      print("@@ Succeeded with list creation at partition " + index)
      partition.map(record => tenMillionList.size + "_" + record)
    })
    println(mappedNumbersRDD.collect().toList)
  }

}

File: Chapter04/Exercise24_04/Exercise24_04.py
from typing import List, Iterable
from pyspark.rdd import RDD
from pyspark.sql import SparkSession
import time


def partition_function(index: int, partition: Iterable):
    print('@@ Starting with list creation at partition ' + str(index))
    four_dozen_m_list: List[int] = []
    for number in range(0, 48000000):
        four_dozen_m_list.append(-1)
    print('@@ Succeeded with list creation at partition ' + str(index))
    time.sleep(30)
    return iter([str(len(four_dozen_m_list)) + '_' + str(record) for record in partition])


if __name__ == "__main__":
    session: SparkSession = SparkSession.builder \
        .master('local[{}]'.format(4)) \
        .appName("Memory Limits") \
        .getOrCreate()

    numbers_rdd: RDD = session.sparkContext.range(0, 10)
    mapped_numbers_rdd = numbers_rdd.mapPartitionsWithIndex(partition_function)
    print(mapped_numbers_rdd.collect())
